File: Hyperiso/Hyperiso/examples_python/Core/database_writer_example.py
"""Export the initialized Core database to JSON, YAML and FLHA files."""

import argparse
from pathlib import Path

import pyhyperiso
from pyhyperiso.Common import Model, ParameterType
from pyhyperiso.Core import DatabaseWriter, HyperisoConfig, HyperisoMaster
from pyhyperiso.core.Common.ParamId import ParamId


def main() -> None:
    """Initialize HyperIso and demonstrate full and selective database exports."""
    default_lha = Path(pyhyperiso.__file__).resolve().parent / "assets" / "lha" / "si_input.flha"
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--lha", type=Path, default=default_lha, help="Input FLHA/LHA file")
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=Path("database_exports_python"),
        help="Directory receiving the exported database files",
    )
    args = parser.parse_args()

    hyperiso = HyperisoMaster()
    hyperiso.init(
        lha_file=str(args.lha),
        config=HyperisoConfig(model=Model.SM),
    )

    output_dir = args.output_dir
    output_dir.mkdir(parents=True, exist_ok=True)

    writer = DatabaseWriter()

    # Export the complete initialized Core database.
    writer.write(output_dir / "database.json")
    writer.write(output_dir / "database.yaml")
    writer.write(output_dir / "database.flha")

    # Export a block subset.
    writer.write_blocks(
        output_dir / "sm_inputs.yaml",
        ["SMINPUTS", "MASS"],
    )

    # Export selected block/id entries.
    writer.write_parameters(
        output_dir / "selected_parameters.json",
        [
            ParamId(type=ParameterType.SM, block="MASS", code=25),
            ParamId(type=ParameterType.SM, block="SMINPUTS", code=3),
        ],
    )

    print(f"Database exports written to {output_dir.resolve()}")


if __name__ == "__main__":
    main()

File: Hyperiso/Hyperiso/examples_cpp/Core/database_writer_example.cpp
#include <filesystem>
#include <iostream>
#include <vector>

#include "Config.h"
#include "FileWriter.h"
#include "HyperisoMaster.h"
#include "Include.h"
#include "ParamID.h"

int main(int argc, char** argv) {
    HyperisoConfig config;
    config.model = Model::SM;
    config.flags[ExternalFlag::IS_LHA_SPECTRUM] = false;
    config.flags[ExternalFlag::HAS_WILSON_INPUT] = false;
    config.flags[ExternalFlag::HAS_TH_OBSERVABLE_INPUT] = false;
    config.flags[ExternalFlag::HYP_AS_SM_MARTY] = true;

    const std::filesystem::path lha_path =
        argc > 1 ? std::filesystem::path(argv[1])
                 : std::filesystem::path("Assets/lha/si_input.flha");
    const std::filesystem::path output_dir =
        argc > 2 ? std::filesystem::path(argv[2])
                 : std::filesystem::path("database_exports_cpp");

    HyperisoMaster hyperiso;
    hyperiso.init(lha_path.string(), config);

    std::filesystem::create_directories(output_dir);

    FileWriter writer;

    // Export the complete initialized Core database in the three supported
    // serialization families.
    writer.write((output_dir / "database.json").string());
    writer.write((output_dir / "database.yaml").string());
    writer.write((output_dir / "database.flha").string());

    // Export a block subset.
    writer.write_blocks(
        (output_dir / "sm_inputs.yaml").string(),
        {BlockName("SMINPUTS"), BlockName("MASS")}
    );

    // Export selected block/id entries.
    writer.write_parameters(
        (output_dir / "selected_parameters.json").string(),
        {
            ParamId(ParameterType::SM, BlockName("MASS"), LhaID(25)),
            ParamId(ParameterType::SM, BlockName("SMINPUTS"), LhaID(3)),
        }
    );

    std::cout << "Database exports written to " << output_dir << '\n';
    return 0;
}
